pkg/net/net_linux.go

```diff
@@ -103,99 +83,41 @@ func netDeviceMacAddress(paths *linuxpath.Paths, dev string) string {
 	return strings.TrimSpace(string(contents))
 }
 
-func ethtoolInstalled() bool {
-	_, err := exec.LookPath("ethtool")
-	return err == nil
-}
-
-func (n *NIC) netDeviceParseEthtool(ctx *context.Context, dev string) {
-	var out bytes.Buffer
-	path, _ := exec.LookPath("ethtool")
-
-	// Get auto-negotiation and pause-frame-use capabilities from "ethtool" (with no options)
-	// Populate Speed, Duplex, SupportedLinkModes, SupportedPorts, SupportedFECModes,
-	// AdvertisedLinkModes, and AdvertisedFECModes attributes from "ethtool" output.
-	cmd := exec.Command(path, dev)
-	cmd.Stdout = &out
-	err := cmd.Run()
-	if err == nil {
-		m := parseNicAttrEthtool(&out)
-		n.Capabilities = append(n.Capabilities, autoNegCap(m))
-		n.Capabilities = append(n.Capabilities, pauseFrameUseCap(m))
-
-		// Update NIC Attributes with ethtool output
-		n.Speed = strings.Join(m["Speed"], "")
-		n.Duplex = strings.Join(m["Duplex"], "")
-		n.SupportedLinkModes = m["Supported link modes"]
-		n.SupportedPorts = m["Supported ports"]
-		n.SupportedFECModes = m["Supported FEC modes"]
-		n.AdvertisedLinkModes = m["Advertised link modes"]
-		n.AdvertisedFECModes = m["Advertised FEC modes"]
-	} else {
-		msg := fmt.Sprintf("could not grab NIC link info for %s: %s", dev, err)
-		ctx.Warn(msg)
+func netDeviceCapabilities(ctx *context.Context, dev string) []*NICCapability {
+	ethHandle, err := ethtool.NewEthtool()
+	if err != nil {
+		ctx.Warn("failed to create ethtool instance: %v", err)
+		return []*NICCapability{}
 	}
-
-	// Get all other capabilities from "ethtool -k"
-	cmd = exec.Command(path, "-k", dev)
-	cmd.Stdout = &out
-	err = cmd.Run()
-	if err == nil {
-		// The out variable will now contain something that looks like the
-		// following.
-		//
-		// Features for enp58s0f1:
-		// rx-checksumming: on
-		// tx-checksumming: off
-		//     tx-checksum-ipv4: off
-		//     tx-checksum-ip-generic: off [fixed]
-		//     tx-checksum-ipv6: off
-		//     tx-checksum-fcoe-crc: off [fixed]
-		//     tx-checksum-sctp: off [fixed]
-		// scatter-gather: off
-		//     tx-scatter-gather: off
-		//     tx-scatter-gather-fraglist: off [fixed]
-		// tcp-segmentation-offload: off
-		//     tx-tcp-segmentation: off
-		//     tx-tcp-ecn-segmentation: off [fixed]
-		//     tx-tcp-mangleid-segmentation: off
-		//     tx-tcp6-segmentation: off
```

Comment:
Are the capability strings *exactly* the same with the ethtool library as they were with the raw parsing of the `ethtool` binary?

Reply:
right, pretty bad oversight on my side. Let me add a test which ensure this. On the code side, if this is not already the case, we can add a thin translation layer.

Reply:
@ffromani I was only making sure :) it may be that the ethtool library does use the exact same strings. but if not, we'll need to add a translation table to ensure backwards compat.

Reply:
OK, so the ethtool program gets the identifiers from the kernel using the ioctl interface: https://kernel.googlesource.com/pub/scm/network/ethtool/ethtool/+/refs/heads/ethtool-3.4.y/ethtool.c#1310 https://docs.kernel.org/networking/ethtool-netlink.html#strset-get

And so it does the ethtool package: https://github.com/safchain/ethtool/blob/master/ethtool.go#L733

Hence the names should be exactly the same, _unless_ it is the `ethtool` program which does some translation, which I haven't checked yet. On it.